src/inner_product_proof.rs

```diff
@@ -66,16 +62,68 @@ impl InnerProductProof {
 
         transcript.innerproduct_domain_sep(n as u64);
 
-        // XXX save these scalar mults by unrolling them into the
-        // first iteration of the loop below
-        for (H_i, h_i) in H.iter_mut().zip(Hprime_factors.into_iter()) {
-            *H_i = (&*H_i) * h_i.borrow();
-        }
-
         let lg_n = n.next_power_of_two().trailing_zeros() as usize;
         let mut L_vec = Vec::with_capacity(lg_n);
         let mut R_vec = Vec::with_capacity(lg_n);
 
+        // If it's the first iteration, unroll the Hprime = H*y_inv scalar mults
+        // into multiscalar muls, for performance.
+        if n != 1 {
+            n = n / 2;
+            let (a_L, a_R) = a.split_at_mut(n);
+            let (b_L, b_R) = b.split_at_mut(n);
+            let (G_L, G_R) = G.split_at_mut(n);
+            let (H_L, H_R) = H.split_at_mut(n);
+
+            let c_L = inner_product(&a_L, &b_R);
+            let c_R = inner_product(&a_R, &b_L);
+
+            let h_l: Vec<_> = b_R
+                .iter()
+                .zip(Hprime_factors[0..n].into_iter())
+                .map(|(b_R_i, y_i)| b_R_i * y_i)
+                .collect();
+            let h_r: Vec<_> = b_L
+                .iter()
+                .zip(Hprime_factors[n..2 * n].into_iter())
+                .map(|(b_L_i, y_i)| b_L_i * y_i)
+                .collect();
```

Comment:
Instead of collecting these, I think if you just do
```
let h_r = b_L
                .iter()
                .zip(Hprime_factors[n..2 * n].into_iter())
                .map(|(b_L_i, y_i)| b_L_i * y_i);
```
you can pass `h_l`, `h_r` into the multiscalar mul directly.